Studies/source/MvaTraining.cxx

```diff
@@ -162,43 +354,139 @@ class MVATraining {
             }
         }
+        vars->UploadEvents();
+        vars->loader->PrepareTrainingAndTestTree( "","", "SplitMode=Random" );
+        MvaOptionCollection options = mva_setup.CreateOptionCollection(true);
+
+        const Grid_ND grid(options.GetPositionLimits());
+        std::cout<<"grid"<<std::endl;
+        std::map<std::string, std::string> methods;
+        for(const auto& point : grid){
+            methods[options.GetName(point)+"_"+std::to_string(args.seed())] = options.GetConfigString(point);
+            std::cout << options.GetName(point)+"_"+std::to_string(args.seed()) << "    " << options.GetConfigString(point) << std::endl;
+        }
+        std::cout << methods.size() << " metodi" << std::endl;
 
-        vars.UploadEvents();
-        dataloader->PrepareTrainingAndTestTree( "","", "SplitMode=Random" );
-
-        factory->BookMethod( dataloader.get(), TMVA::Types::kBDT, "BDTG",
-                             "!H:!V:NTrees=1000:MinNodeSize=2.5%:BoostType=Grad:Shrinkage=0.10:UseBaggedBoost:BaggedSampleFraction=0.5:nCuts=20:MaxDepth=6");
+        for(const auto& m : methods)
+            factory->BookMethod(vars->loader.get(), TMVA::Types::kBDT, m.first, m.second);
 
+        TrainAllMethods(factory);
+//        factory->TrainAllMethods();
+        factory->TestAllMethods();
+        factory->EvaluateAllMethods();
 
-        factory->BookMethod( dataloader.get(), TMVA::Types::kBDT, "BDTD_2","!H:!V:NTrees=200:MaxDepth=3:MinNodeSize=2.5%:"
-                                        "nCuts=-1:BoostType=Bagging:UseBaggedBoost=True:BaggedSampleFraction=0.6:UseRandomisedTrees=True:"
-                                        "UseNvars=2:SeparationType=GiniIndex:UseYesNoLeaf:VarTransform=Decorrelate" );
+        for(const auto& m : methods){
+            TString weightfile = ("mydataloader/weights/myFactory_"+m.first+".weights.xml");
+            vars->reader->BookMVA(m.first, weightfile);
+        }
 
+        auto importance = RankingImportanceVariables(grid, options);
+        std::cout<<"importance"<<std::endl;
+        auto position = RankingPoisitionVariables(importance);
+        std::cout<<"position"<<std::endl;
+
+        std::map<std::string, std::map<SampleId, std::map<size_t, std::vector<double>>>> evaluation;
+        std::map<std::string, std::map<SampleId, std::shared_ptr<TH1D>>> outputBDT;
+        for(const auto& m : methods){
+            outputBDT[m.first][mass_tot] = std::make_shared<TH1D>(("Signal_output_"+m.first).c_str(),("Signal_output_"+m.first).c_str(), 200, -1, 1);
+            outputBDT[m.first][bkg] = std::make_shared<TH1D>(("Bkg_output_"+m.first).c_str(),("Bkg_output_"+m.first).c_str(), 200, -1, 1);
+            for(const auto& type_entry : vars->data){
+                auto sample_entry = type_entry.second;
+                for(const auto& entry : sample_entry){
+                    evaluation[m.first][entry.first][type_entry.first] = vars->EvaluateForAllEvents(m.first, type_entry.first, entry.first);
```

Comment:
It looks correct, but to improve readability here and in the next line, better to define
auto& current_evolution = evaluation[m.first][entry.first][type_entry.first]
similarly, for evaluation[m.first][mass_tot][type_entry.first]